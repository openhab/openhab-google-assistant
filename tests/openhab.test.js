
const OpenHAB = require('../functions/openhab.js').OpenHAB;

describe('Test SYNC', () => {
  test('light items', async () => {
    const items = [
      {
        "link": "https://openhab.example.org/rest/items/MySwitch",
        "state": "OFF",
        "editable": false,
        "type": "Switch",
        "name": "MySwitch",
        "label": "SwitchLight",
        "tags": [
          "Lighting"
        ],
        "groupNames": []
      },
      {
        "link": "https://openhab.example.org/rest/items/MyDimmer",
        "state": "0",
        "editable": false,
        "type": "Dimmer",
        "name": "MyDimmer",
        "label": "DimmLight",
        "tags": [
          "Lighting"
        ],
        "groupNames": []
      },
      {
        "link": "https://openhab.example.org/rest/items/MyLight",
        "state": "0,0,0",
        "editable": false,
        "type": "Color",
        "name": "MyLight",
        "label": "ColorLight",
        "tags": [
          "Lighting"
        ],
        "groupNames": []
      },
      {
        "members": [],
        "link": "https://openhab.example.org/rest/items/MyLightGroup",
        "state": "NULL",
        "editable": false,
        "type": "Group",
        "groupType": "Switch",
        "name": "MyLightGroup",
        "label": "GroupLight",
        "tags": [
          "Lighting"
        ],
        "groupNames": []
      },
      {
        "members": [],
        "link": "https://openhab.example.org/rest/items/MyDimmerGroup",
        "state": "NULL",
        "editable": false,
        "type": "Group",
        "groupType": "Dimmer",
        "name": "MyDimmerGroup",
        "label": "GroupDimmer",
        "tags": [
          "Lighting"
        ],
        "groupNames": []
      },
      {
        "members": [],
        "link": "https://openhab.example.org/rest/items/MyColorGroup",
        "state": "NULL",
        "editable": false,
        "type": "Group",
        "groupType": "Color",
        "name": "MyColorGroup",
        "label": "GroupColor",
        "tags": [
          "Lighting"
        ],
        "groupNames": []
      }
    ];

    const getItemsMock = jest.fn();
    getItemsMock.mockReturnValue(Promise.resolve(items));

    const apiHandler = {
      getItems: getItemsMock
    };

    const payload = await new OpenHAB(apiHandler).handleSync();

    expect(payload).toMatchSnapshot();
  });
});

describe('Test QUERY', () => {
  test('single light item', async () => {
    const item =
    {
      "link": "https://openhab.example.org/rest/items/MySwitch",
      "state": "OFF",
      "editable": false,
      "type": "Switch",
      "name": "MySwitch",
      "label": "SwitchLight",
      "tags": [
        "Lighting"
      ],
      "groupNames": []
    };

    const getItemMock = jest.fn();
    getItemMock.mockReturnValue(Promise.resolve(item));

    const apiHandler = {
      getItem: getItemMock
    };

    const payload = await new OpenHAB(apiHandler).handleQuery([{
      "id": "MySwitch"
    }]);

    expect(payload).toStrictEqual({
      "devices": {
        "MySwitch": {
          "on": false,
          "online": true,
        },
      },
    });
  });

  test('multiple light items', async () => {
    const item1 =
    {
      "link": "https://openhab.example.org/rest/items/MySwitch",
      "state": "OFF",
      "editable": false,
      "type": "Switch",
      "name": "MySwitch",
      "label": "SwitchLight",
      "tags": [
        "Lighting"
      ],
      "groupNames": []
    };

    const item2 =
    {
      "link": "https://openhab.example.org/rest/items/MyDimmer",
      "state": "20",
      "editable": false,
      "type": "Dimmer",
      "name": "MyDimmer",
      "label": "DimmerLight",
      "tags": [
        "Lighting"
      ],
      "groupNames": []
    };

    const getItemMock = jest.fn();
    getItemMock.mockReturnValueOnce(Promise.resolve(item1))
      .mockReturnValueOnce(Promise.resolve(item2));

    const apiHandler = {
      getItem: getItemMock
    };

    const payload = await new OpenHAB(apiHandler).handleQuery([{
      "id": "MySwitch"
    }, {
      "id": "MyDimmer"
    }]);

    expect(payload).toStrictEqual({
      "devices": {
        "MySwitch": {
          "on": false,
          "online": true,
        },
        "MyDimmer": {
          "on": true,
          "brightness": 20,
          "online": true,
        },
      },
    });
  });
});
