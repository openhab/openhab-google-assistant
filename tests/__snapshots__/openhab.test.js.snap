// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test SYNC light items 1`] = `
Object {
  "devices": Array [
    Object {
      "attributes": Object {},
      "customData": Object {
        "itemTag": "Lighting",
        "itemType": "Switch",
        "openhabVersion": "2.4.0",
      },
      "deviceInfo": Object {
        "hwVersion": "2.4.0",
        "manufacturer": "openHAB",
        "model": "Lighting",
        "swVersion": "2.4.0",
      },
      "id": "MySwitch",
      "name": Object {
        "name": "SwitchLight",
      },
      "traits": Array [
        "action.devices.traits.OnOff",
      ],
      "type": "action.devices.types.LIGHT",
      "willReportState": false,
    },
    Object {
      "attributes": Object {},
      "customData": Object {
        "itemTag": "Lighting",
        "itemType": "Dimmer",
        "openhabVersion": "2.4.0",
      },
      "deviceInfo": Object {
        "hwVersion": "2.4.0",
        "manufacturer": "openHAB",
        "model": "Lighting",
        "swVersion": "2.4.0",
      },
      "id": "MyDimmer",
      "name": Object {
        "name": "DimmLight",
      },
      "traits": Array [
        "action.devices.traits.OnOff",
        "action.devices.traits.Brightness",
      ],
      "type": "action.devices.types.LIGHT",
      "willReportState": false,
    },
    Object {
      "attributes": Object {
        "colorModel": "hsv",
      },
      "customData": Object {
        "itemTag": "Lighting",
        "itemType": "Color",
        "openhabVersion": "2.4.0",
      },
      "deviceInfo": Object {
        "hwVersion": "2.4.0",
        "manufacturer": "openHAB",
        "model": "Lighting",
        "swVersion": "2.4.0",
      },
      "id": "MyLight",
      "name": Object {
        "name": "ColorLight",
      },
      "traits": Array [
        "action.devices.traits.OnOff",
        "action.devices.traits.Brightness",
        "action.devices.traits.ColorSetting",
      ],
      "type": "action.devices.types.LIGHT",
      "willReportState": false,
    },
    Object {
      "attributes": Object {},
      "customData": Object {
        "itemTag": "Lighting",
        "itemType": "Group",
        "openhabVersion": "2.4.0",
      },
      "deviceInfo": Object {
        "hwVersion": "2.4.0",
        "manufacturer": "openHAB",
        "model": "Lighting",
        "swVersion": "2.4.0",
      },
      "id": "MyLightGroup",
      "name": Object {
        "name": "GroupLight",
      },
      "traits": Array [
        "action.devices.traits.OnOff",
      ],
      "type": "action.devices.types.LIGHT",
      "willReportState": false,
    },
    Object {
      "attributes": Object {},
      "customData": Object {
        "itemTag": "Lighting",
        "itemType": "Group",
        "openhabVersion": "2.4.0",
      },
      "deviceInfo": Object {
        "hwVersion": "2.4.0",
        "manufacturer": "openHAB",
        "model": "Lighting",
        "swVersion": "2.4.0",
      },
      "id": "MyDimmerGroup",
      "name": Object {
        "name": "GroupDimmer",
      },
      "traits": Array [
        "action.devices.traits.OnOff",
        "action.devices.traits.Brightness",
      ],
      "type": "action.devices.types.LIGHT",
      "willReportState": false,
    },
    Object {
      "attributes": Object {
        "colorModel": "hsv",
      },
      "customData": Object {
        "itemTag": "Lighting",
        "itemType": "Group",
        "openhabVersion": "2.4.0",
      },
      "deviceInfo": Object {
        "hwVersion": "2.4.0",
        "manufacturer": "openHAB",
        "model": "Lighting",
        "swVersion": "2.4.0",
      },
      "id": "MyColorGroup",
      "name": Object {
        "name": "GroupColor",
      },
      "traits": Array [
        "action.devices.traits.OnOff",
        "action.devices.traits.Brightness",
        "action.devices.traits.ColorSetting",
      ],
      "type": "action.devices.types.LIGHT",
      "willReportState": false,
    },
  ],
}
`;
